midend/replaceSelectRange.cpp

```diff
@@ -0,0 +1,175 @@
+/*
+* Copyright 2020, MNK Labs & Consulting
+* http://mnkcg.com
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+*    http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*
+*/
+
+#include "replaceSelectRange.h"
+
+namespace P4 {
+
+static int trailingZeros(big_int n, int width) {
+    int zeros = 0;
+
+    while (n > 0 && (n & 1) == 0) {
+        zeros += 1;
+        n >>= 1;
+    }
+
+    return (zeros < width) ? zeros : width;
+}
+
+bool DoReplaceSelectRange::checkRange(const IR::Range* range) {
```

Comment:
This function is only called once, I think it is better to inline it. Then your code will become much more compact.
E.g.: 
```
auto l = range->left->to<IR::Constant>();
if (!l) {  ::error("..."); 
```

Reply:
checkRange() is called twice because rangeToMasks() is called twice.

Reply:
I can see exactly 1 line in your code with a call to checkRange().

Reply:
Yes, but rangeToMasks() is called twice and thus checkRange() ends up getting calls twice.

Reply:
I think Mihai's comment is that regardless of the number of times that checkRange() is called at execution time, it is only called from one place in the source code, so in his opinion, the code would read nicer if there were no separate function called checkRange(), and that its body was simply incorporated into the code where checkRange() is called from.  It isn't a correctness concern, but a code readability style suggestion.

Reply:
Thanks for clarifying, @jafingerhut.  I will change the code.

Reply:
My viewpoint was that the code could modular by having a different checmRange() method and keeping the computing guts in a different rangeToMasks().  I have no strong opinion and can do whatever we go with.